hedera-node/src/main/java/com/hedera/services/context/primitives/SignedStateViewFactory.java

```diff
@@ -0,0 +1,95 @@
+package com.hedera.services.context.primitives;
+
+import com.hedera.services.ServicesState;
+import com.hedera.services.context.ImmutableStateChildren;
+import com.hedera.services.context.MutableStateChildren;
+import com.hedera.services.exceptions.NoValidSignedStateException;
+import com.hedera.services.state.migration.StateVersions;
+import com.swirlds.common.AutoCloseableWrapper;
+import com.swirlds.common.Platform;
+
+import java.util.Optional;
+
+/**
+ * Factory that provides view of last completed signed state and its state children.
+ * Fetches the latest signed state from platform for each call to get/update state children.
+ */
+public class SignedStateViewFactory {
+	private final Platform platform;
+
+	public SignedStateViewFactory(Platform platform) {
+		this.platform = platform;
+	}
+
+	/**
+	 * Try to update the mutable state children with the children from the signed state.
+	 * Throws {@code NoValidSignedStateException} when the provided state is invalid.
+	 *
+	 * @param children
+	 * 		mutable children provided
+	 * @throws NoValidSignedStateException
+	 */
+	public void tryToUpdate(MutableStateChildren children) throws NoValidSignedStateException {
+		final var state = getLatestSignedState();
```

Comment:
Ah, I just remember that we are supposed to only use the state returned from `platform.getLastCompleteSwirldState()` **while inside the `try`-with-resources block**.

So we prob want to tweak your `getLatestSignedState()` method to be,
```
private void doWithLatest(Consumer<ServicesState> action) throws NoValidSignedStateException {
    try (final AutoCloseableWrapper<ServicesState> wrapper = platform.getLastCompleteSwirldState()) {
        final var signedState = wrapper.get();
        if (!isValid(signedState)) {
            throw new NoValidSignedStateException();
        }
        action.accept(signedState);
    }
}
```

And then `tryToUpdate()` becomes,
```
public void tryToUpdate(MutableStateChildren children) throws NoValidSignedStateException {
    doWithLatest(state -> children.updateFromSigned(state, state.getTimeOfLastHandledTxn()));
}
```

And `tryToGet()` becomes,
```
public Optional<StateChildren> tryToGetLatestSignedChildren() {
    final var ref = new AtomicReference<StateChildren>();
    try {
        doWithLatest(state -> ref.set(new ImmutableStateChildren(state)));
        return ref.get();     
    } catch (NoValidSignedStateException ignore) {
        return Optional.empty();
    }
}
```

Reply:
Fixed ! Thanks !